src/eval_env.cc

```diff
@@ -63,6 +63,16 @@ const EvalString* Rule::GetBinding(const string& key) const {
   return &i->second;
 }
 
+std::unique_ptr<Rule> Rule::Phony() {
+  auto rule = std::unique_ptr<Rule>(new Rule("phony"));
+  rule->phony_ = true;
+  return rule;
+}
+
+bool Rule::IsPhony() const {
+  return phony_;
```

Comment:
nit: Can we just store the address of this Rule in `State::phony_rule_` and implement `State::IsPhony(const Rule* rule) { return rule == phony_rule_; }` instead? This would avoid adding 8 bytes to each Rule instance due to padding.

Reply:
you mean like it was before? That doesn't work anymore since then the Rule would be deleted twice in `std::map<std::string, std::unique_ptr<const Rule>> rules_;`, wouldn't it?

Reply:
Note necessarily with something like:

```
// state.h
struct State {
  ...
  const Rule* phony_rule() const { return phony_rule_; }
  
  Rule* phony_rule_ = nullptr;
  ...
};

// state.cc
State::State() {
  ...
  // Add the phony rule to it.
  phony_rule_ = new Rule("phony");
  bindings_.AddRule(phony_rule_);  // passes ownership to bindings_.
}
```

The caveat is that you need a State pointer whenever your want to implement `rule->IsPhony()`, but this is a minor modification to the rest of the code.

Frankly, this could go in a different PR. The memory leak fix is more important, so feel free to submit if you want.